add_overlay extended

mentos987 opened the pull request:
First of. I am a novice to github and c++, no doubt  there will be rewrites needed. 

This is done in order to extend image placement functions to allow for: animations, multi-hex images and drawing layer inputs. All optional inputs.

This will give the content creators a lot more freedom in what they want their images to do.

Other changes:
Bug fix were cutting an image with an "image::HEXED" caused the image to darken.
Image removal is changed to allow for multi-hex removal.
Image removal also gets optional input to allow for a partial string match rather than a full one. This makes it simpler to remove an image when you don't know its full string, something that can happen with a lot of IPF modifications. (This is not needed for the rest of this code to work)

This is tested on Windows 10 and seems to work fine. No issues detected.

Call examples:

    //wml

    //add multihex animation
        [item]
	        x,y=105,158
	        image=data/add-ons/Animated_Weather_and_Scenery/images/weather/rain_heavy/00[01~20].png
	        multihex=yes
        [/item]
    
    //remove multihex animation from above
        [remove_item]
	        x,y=105,158
	        image=weather/rain_heavy
	        do_partial_string_match=yes
        [/remove_item]
    
    //lua
    
    //add great continent map on top of in-game terrain (see image below)
    //add " drawing_layer = "arrows" " in there and the image will be drawn at the arrow level (above the units)
        wesnoth.interface.add_hex_overlay(68, 57, { image = "data/add-ons/Animated_Weather_and_Scenery/images/great_continent.png", multihex = "yes"})

    //remove
        wesnoth.interface.remove_hex_overlay(68, 57, "ages/great_continent", true)

<img width="687" height="429" alt="image" src="https://github.com/user-attachments/assets/0cd693f3-4599-457f-b04a-c20ef1b95c17" />


Edit, this has CRLF line endings, I'm guessing if should be LF. I'll change over later.

Edit 2, it also occurs to me now that a multi-hex image may get 1-hex-size children that are fully transparent. I will weed these out next.

Edit 3, When removing the children overlays we accidentally clear the entire child hex of all overlays, not just the child. I will change this.

Edit 4, no need to recall make_overlay_crop() for every frame, 1 time per hex is enough. Will rework.

Edit 5, also missed to enable "drawing_layer" as input from WML code (only Lua atm). Will add.

mentos987 committed:
add_overlay extended

hladkyjakub commented:
It may be worth noticing that unit tests will fail, unless special care.
At the same time the special care should be given to it only once rest of the code is reviewed and submited.

Wedge009 commented:
> First of. I am a novice to github and c++, no doubt there will be rewrites needed.

Good on you for giving it a go anyway.

> Edit, this has CRLF line endings, I'm guessing if should be LF. I'll change over later.

I think if you configured your local git properly, this shouldn't be an issue. GitHub diff seems to handle this fine, anyway.

I suggest if you're adding commentary as you make progress in your work, you just add another comment rather than make edits, unless you're just making corrections to your previous statements.

ProditorMagnus commented:
I do not see use case for do_partial_string_match. Add name to item, and remove_item by that name.

mentos987 commented:
> > Other changes:
> > Bug fix were cutting an image with an "image::HEXED" caused the image to darken.
> > Image removal is changed to allow for multi-hex removal.
> > Image removal also gets optional input to allow for a partial string match rather than a full one. This makes it simpler to remove an image when you don't know its full string, something that can happen with a lot of IPF modifications. (This is not needed for the rest of this code to work)
> 
> Do not mix multiple independent changes into the same commit. All those changes should be separate commits or even separate PRs. That will make it much easier and faster to review and potentially merge.

The problem is that only the "partial string match" is independent, and barely at that. The "image::HEXED" fix is needed for the rest of this to work.

soliton- commented:
> The problem is that only the "partial string match" is independent, and barely at that.

If you think it is useless without other changes here then make it its own commit in this PR.

> The "image::HEXED" fix is needed for the rest of this to work and so is the multi-hex removal.

If the bug fix is truely not relevant for anything else (how is it even a bug fix?) then put it in separate commit in this PR (before a commit that uses what is fixed). If multi-hex removal is not independent at all then it's just confusing to list it as "other changes".

Btw, none of your explanations here are in the commit message. See https://wiki.wesnoth.org/DeveloperGuide#Commits for guidelines on how to write good commits that are easy to understand and review.

mentos987 commented:
> If you think it is useless without other changes here then make it its own commit in this PR.

I'll try to in future.

> If the bug fix is truly not relevant for anything else (how is it even a bug fix?) then put it in separate commit in this PR (before a commit that uses what is fixed). If multi-hex removal is not independent at all then it's just confusing to list it as "other changes".

It is a general bug. I'll keep it in mind for future.

> Btw, none of your explanations here are in the commit message.

I'll have a look.

mentos987 committed:
Several fixes

mentos987 commented:
I have fixed a few things but some things remains to be done.
Changes:

1. "multi-hex image may get 1-hex-size children that are fully transparent. I will weed these out next." -- I did not fix this since a solution for it is done at a later stage at load_image_sub_file function in pictures.cpp
2. Reworked make_overlay_crop function and prevented it from repeating unneeded calls
3. "drawing_layer" is now usable as input from WML [item] call 
4. Reworked image removal, it was not correctly removing the children of a multi-hex image before.

Remaining things to do:

1. Failing some tests, at first glance I cant make sense of it. Will have to revisit.
2. Replacement of the do_partial_string_match boolean for an string match function. Waiting for clarification of how it should work.
3. Revisit drawing_layer names to make sure they're good for a public API. Leaving this for others.
4. May revisit how drawing_layer input should look like.

mentos987 commented:
> Btw, none of your explanations here are in the commit message.

Missed doing this again, I did not realise that the commits here are the text I make on my desktop github.

mentos987 committed:
TOD fix - enabled for animations.

mentos987 committed:
Fix for the testing failures, also TOD fix for animations

mentos987 commented:
Restarting this PR with https://github.com/wesnoth/wesnoth/pull/10518 in order to simplify reviews and feedback.
